File: lib/scoring_rules.dart
import 'dart:math';

import 'package:sagrada/game.dart';
import 'package:flutter_gen/gen_l10n/app_localizations.dart';

class ScoringResult {
  final int score;
  final List<List<bool>> mask;
  final String calculation;

  ScoringResult(this.score, this.mask, this.calculation);
}

abstract class ScoringRule {
  ScoringResult getScore(Board board);
  String getTranslation(AppLocalizations l10n);
}

const times = '×';

class SumColor extends ScoringRule {
  final Color color;

  SumColor(this.color);

  @override
  ScoringResult getScore(Board board) {
    final mask = board.createMask((dice, i, j) => dice?.color == color);
    final numbers = board.diceAtMask(mask).map((dice) => dice!.number);
    final score = numbers.reduce((a, b) => a + b);

    return ScoringResult(score, mask, numbers.join(" + "));
  }

  @override
  String getTranslation(AppLocalizations l10n) {
    return l10n.privateGoal(color.getTranslation(l10n));
  }
}

class BlankPenalty extends ScoringRule {
  @override
  ScoringResult getScore(Board board) {
    final mask = board.createMask((dice, i, j) => dice == null);
    final numBlanks = board.diceAtMask(mask).length;
    final score = numBlanks * -1;

    return ScoringResult(score, mask, "$numBlanks $times (-1)");
  }

  @override
  String getTranslation(AppLocalizations l10n) {
    return l10n.blankPenalty;
  }
}

typedef SetClassifier = bool Function(Dice?);

abstract class SetBasedScoringRule<T> extends ScoringRule {
  int scorePerSet;
  List<SetClassifier> diceClassifiers;

  SetBasedScoringRule(this.scorePerSet, this.diceClassifiers);

  @override
  ScoringResult getScore(Board board) {
    final itemCounts = List.filled(diceClassifiers.length, 0);

    board.forEachDice((dice, i, j) {
      for (int k = 0; k < diceClassifiers.length; k++) {
        if (diceClassifiers[k](dice)) {
          itemCounts[k]++;
        }
      }
    });

    final lowestItemCount = itemCounts.reduce(min);

    final leastFrequentClassesIndices = itemCounts.indexed
        .where((el) => el.$2 == lowestItemCount)
        .map((el) => el.$1);

    final mask = board.createMask((dice, i, j) {
      return leastFrequentClassesIndices.any((k) => diceClassifiers[k](dice));
    });

    final score = lowestItemCount * scorePerSet;

    return ScoringResult(
        score, mask, "min(${itemCounts.join(', ')}) $times $scorePerSet");
  }
}

class LightShades extends SetBasedScoringRule {
  LightShades()
      : super(2, [
          (dice) => dice?.number == 1,
          (dice) => dice?.number == 2,
        ]);

  @override
  String getTranslation(AppLocalizations l10n) {
    return l10n.lightShades;
  }
}

class MediumShades extends SetBasedScoringRule {
  MediumShades()
      : super(2, [
          (dice) => dice?.number == 3,
          (dice) => dice?.number == 4,
        ]);

  @override
  String getTranslation(AppLocalizations l10n) {
    return l10n.mediumShades;
  }
}

class DeepShades extends SetBasedScoringRule {
  DeepShades()
      : super(2, [
          (dice) => dice?.number == 5,
          (dice) => dice?.number == 6,
        ]);

  @override
  String getTranslation(AppLocalizations l10n) {
    return l10n.deepShades;
  }
}

class ColorVariety extends SetBasedScoringRule {
  ColorVariety()
      : super(
            4,
            Color.values
                .map((color) => (dice) => dice?.color == color)
                .toList());

  @override
  String getTranslation(AppLocalizations l10n) {
    return l10n.colorVariety;
  }
}

class ShadeVariety extends SetBasedScoringRule {
  ShadeVariety()
      : super(
            5,
            [1, 2, 3, 4, 5, 6]
                .map((number) => (dice) => dice?.number == number)
                .toList());

  @override
  String getTranslation(AppLocalizations l10n) {
    return l10n.shadeVariety;
  }
}

abstract class ColumnVariety<T> extends ScoringRule {
  int scorePerColumn;

  ColumnVariety(this.scorePerColumn);

  T key(Dice dice);

  @override
  ScoringResult getScore(Board board) {
    final satisfyingColumns = <int>{};

    for (int j = 0; j < numColumns; j++) {
      final valuesSeen = <T>{};
      bool isFailed = false;

      for (int i = 0; i < numRows; i++) {
        Dice? dice = board.at(i, j);

        if (dice == null || valuesSeen.contains(key(dice))) {
          isFailed = true;
          break;
        }

        valuesSeen.add(key(dice));
      }

      if (!isFailed) {
        satisfyingColumns.add(j);
      }
    }

    final numSatisfyingColumns = satisfyingColumns.length;
    final mask =
        board.createMask((dice, i, j) => satisfyingColumns.contains(j));
    final score = numSatisfyingColumns * scorePerColumn;

    return ScoringResult(
        score, mask, "$numSatisfyingColumns $times $scorePerColumn");
  }
}

class ColumnColorVariety extends ColumnVariety {
  ColumnColorVariety() : super(5);

  @override
  key(Dice dice) {
    return dice.color;
  }

  @override
  String getTranslation(AppLocalizations l10n) {
    return l10n.columnColorVariety;
  }
}

class ColumnShadeVariety extends ColumnVariety {
  ColumnShadeVariety() : super(4);

  @override
  key(Dice dice) {
    return dice.number;
  }

  @override
  String getTranslation(AppLocalizations l10n) {
    return l10n.columnShadeVariety;
  }
}

abstract class RowVariety<T> extends ScoringRule {
  int scorePerRow;

  RowVariety(this.scorePerRow);

  T key(Dice dice);

  @override
  ScoringResult getScore(Board board) {
    final satisfyingRows = <int>{};

    for (int i = 0; i < numRows; i++) {
      final valuesSeen = <T>{};
      bool isFailed = false;

      for (int j = 0; j < numColumns; j++) {
        Dice? dice = board.at(i, j);

        if (dice == null || valuesSeen.contains(key(dice))) {
          isFailed = true;
          break;
        }

        valuesSeen.add(key(dice));
      }

      if (!isFailed) {
        satisfyingRows.add(i);
      }
    }

    final numSatisfyingColumns = satisfyingRows.length;
    final mask = board.createMask((dice, i, j) => satisfyingRows.contains(i));
    final score = numSatisfyingColumns * scorePerRow;

    return ScoringResult(
        score, mask, "$numSatisfyingColumns $times $scorePerRow");
  }
}

class RowColorVariety extends RowVariety {
  RowColorVariety() : super(6);

  @override
  key(Dice dice) {
    return dice.color;
  }

  @override
  String getTranslation(AppLocalizations l10n) {
    return l10n.rowColorVariety;
  }
}

class RowShadeVariety extends RowVariety {
  RowShadeVariety() : super(5);

  @override
  key(Dice dice) {
    return dice.number;
  }

  @override
  String getTranslation(AppLocalizations l10n) {
    return l10n.rowShadeVariety;
  }
}

class ColorDiagonals extends ScoringRule {
  @override
  ScoringResult getScore(Board board) {
    final mask = getMask(board);
    final countByColors = <Color, int>{};
    board.diceAtMask(mask).forEach((dice) {
      final color = dice!.color;
      countByColors.update(color, (value) => value + 1, ifAbsent: () => 1);
    });
    final score = countByColors.values.reduce((a, b) => a + b);

    return ScoringResult(score, mask, countByColors.values.join(" + "));
  }

  Mask getMask(Board board) {
    final mask =
        List.generate(numRows, (i) => List<bool>.filled(numColumns, false));

    board.forEachDice((dice, i, j) {
      if (dice == null) {
        return;
      }

      final steps = [-1, 1];
      final coordinates = [
        for (final di in steps)
          for (final dj in steps) (i + di, j + dj)
      ].where(validCoordinates);
      final neighbors =
          coordinates.map(((int, int) point) => board.at(point.$1, point.$2));

      mask[i][j] = neighbors.any((neighbor) => neighbor?.color == dice.color);
    });

    return mask;
  }

  @override
  String getTranslation(AppLocalizations l10n) {
    return l10n.colorDiagonals;
  }
}

File: src/game/scoring-rules.ts
import type { Mask, Color, Value, OptionalDice } from "./types";
import { Board, COLORS, NUM_COLUMNS, NUM_ROWS, VALUES } from "./types";

export type ScoringResult = {
  score: number;
  calculation: string;
  mask: Mask;
};

export type ScoringRule = (board: Board) => ScoringResult;

const times = "×";

const checkValue = (value: Value) => (dice: OptionalDice) =>
  dice?.value === value;
const checkColor = (color: Color) => (dice: OptionalDice) =>
  dice?.color === color;

export function blankPenalty(board: Board): ScoringResult {
  const mask = board.createMask((dice) => dice === null);
  const numBlanks = board.diceAtMask(mask).length;
  const penaltyPerBlank = -1;

  return {
    score: numBlanks * penaltyPerBlank,
    calculation: `${numBlanks} ${times} (${penaltyPerBlank})`,
    mask,
  };
}

export function sumColorFactory(color: Color): ScoringRule {
  return (board: Board) => {
    const mask = board.createMask(checkColor(color));
    const numbers = board.diceAtMask(mask).map((dice) => dice?.value ?? 0);
    const score = numbers.reduce((a, b) => a + b, 0);

    return {
      score,
      calculation: numbers.join(" + "),
      mask,
    };
  };
}

function setBasedScoringRuleFactory(
  scorePerSet: number,
  diceClassifiers: ((dice: OptionalDice) => boolean)[],
): ScoringRule {
  return (board: Board) => {
    const itemCounts = diceClassifiers.map(
      (classifier) => board.diceAtMask(board.createMask(classifier)).length,
    );
    const minCount = Math.min(...itemCounts);
    const leastFrequentClassIndices = itemCounts.reduce((acc, count, index) => {
      if (count === minCount) {
        acc.push(index);
      }
      return acc;
    }, [] as number[]);
    const mask = board.createMask((dice) =>
      leastFrequentClassIndices.some((index) => diceClassifiers[index](dice)),
    );
    const score = minCount * scorePerSet;

    return {
      score,
      calculation: `min(${itemCounts.join(", ")}) ${times} ${scorePerSet}`,
      mask,
    };
  };
}

export const lightShades = setBasedScoringRuleFactory(2, [
  checkValue(1),
  checkValue(2),
]);
export const mediumShades = setBasedScoringRuleFactory(2, [
  checkValue(3),
  checkValue(4),
]);
export const deepShades = setBasedScoringRuleFactory(2, [
  checkValue(5),
  checkValue(6),
]);

export const colorVariety = setBasedScoringRuleFactory(
  4,
  COLORS.map(checkColor),
);
export const shadeVariety = setBasedScoringRuleFactory(
  5,
  VALUES.map(checkValue),
);

function columnVarietyFactory<T>(
  scorePerColumn: number,
  diceClassifier: (dice: OptionalDice) => T,
): ScoringRule {
  return (board: Board) => {
    const satisfyingColumns = new Set(
      Array.from({ length: NUM_COLUMNS }, (_, j) => j).filter((j) => {
        const valuesSeen = new Set<T>();

        for (let i = 0; i < NUM_ROWS; i++) {
          const dice = board.at(i, j);

          if (dice === null || valuesSeen.has(diceClassifier(dice))) {
            return false;
          }

          valuesSeen.add(diceClassifier(dice));
        }

        return true;
      }),
    );

    const numSatisfyingColumns = satisfyingColumns.size;
    const mask = board.createMask((dice, i, j) => satisfyingColumns.has(j));
    const score = numSatisfyingColumns * scorePerColumn;

    return {
      score,
      calculation: `${numSatisfyingColumns} ${times} ${scorePerColumn}`,
      mask,
    };
  };
}

export const columnColorVariety = columnVarietyFactory(
  5,
  (dice) => dice?.color,
);
export const columnShadeVariety = columnVarietyFactory(
  4,
  (dice) => dice?.value,
);

function rowVarietyFactory<T>(
  scorePerRow: number,
  diceClassifier: (dice: OptionalDice) => T,
): ScoringRule {
  return (board: Board) => {
    const satisfyingRows = new Set(
      Array.from({ length: NUM_ROWS }, (_, i) => i).filter((i) => {
        const valuesSeen = new Set<T>();

        for (let j = 0; j < NUM_COLUMNS; j++) {
          const dice = board.at(i, j);

          if (dice === null || valuesSeen.has(diceClassifier(dice))) {
            return false;
          }

          valuesSeen.add(diceClassifier(dice));
        }

        return true;
      }),
    );

    const numSatisfyingRows = satisfyingRows.size;
    const mask = board.createMask((_dice, i) => satisfyingRows.has(i));
    const score = numSatisfyingRows * scorePerRow;

    return {
      score,
      calculation: `${numSatisfyingRows} ${times} ${scorePerRow}`,
      mask,
    };
  };
}

export const rowColorVariety = rowVarietyFactory(6, (dice) => dice?.color);
export const rowShadeVariety = rowVarietyFactory(5, (dice) => dice?.value);

export const colorDiagonals = (board: Board) => {
  const mask = board.createMask((dice, i, j) => {
    if (dice === null) {
      return false;
    }

    const neighbors = [
      [i - 1, j - 1],
      [i - 1, j + 1],
      [i + 1, j - 1],
      [i + 1, j + 1],
    ];

    return neighbors
      .filter(([row, column]) => Board.validCoordinates(row, column))
      .some(([row, column]) => {
        const neighbor = board.at(row, column);
        return neighbor !== null && neighbor.color === dice.color;
      });
  });

  const countByColors = board.diceAtMask(mask).reduce(
    (stack, dice) => {
      if (dice === null) {
        return stack;
      }

      stack[dice.color] = (stack[dice.color] || 0) + 1;
      return stack;
    },
    {} as Record<Color, number>,
  );
  const values = Object.values(countByColors);

  return {
    score: values.reduce((a, b) => a + b, 0),
    calculation: values.join(" + "),
    mask,
  };
};
